Persist local-storage in external json-file

ixrock opened the pull request:
- persisting local-storage state in json-files located at `/Users/%USER%/Library/Application Support/LensDev/lens-local-storage/%clusterId%.json`
- refactored `renderer/utils/createStorage` / now it's `observable`
- refactored `Sidebar`, `SidebarItem` and `MainLayout`
- ~~expanding sidebar by clicking arrow-icon (for consistency with compact mode where no sub-menus)~~

close #2166

ixrock committed:
persist local-storage state in external json-file

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix: add immer to package.json

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
renamed compatible storage keys for better migration after merge #2280

Signed-off-by: Roman <ixrock@gmail.com>

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,210 @@
+  };
+
+  private data = observable.box<T>();
+  @observable.ref storage: StorageAdapter<T>;
```
Given that this is is only really assigned in the constructor and that it really shouldn't be changed on the fly. I don't think that this should be either `observable` or `public`. Futhermore, given that this class also stores its `options` wholly this might as well just be a getter.
```suggestion
  private get storage(): StorageAdapter<T> {
    return this.options.storage;
  }
```

ixrock replied:
Updated in #2280

Nokel81 replied:
May I ask why you are updating a follow-up PR with changes requested on this PR?

ixrock replied:
Because it's not targeted to patch-release fix anymore (suggesting to merge #2280 first)

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,210 @@
+    if (observable) this.configureObservable(observable);
 
     return this;
   }
```
Given the above suggestion and that this leaves this function only as an `if (observable)` I would just move that to the `constructor`.

Furthermore, given the rules of how spreading multiple objects into a new object works. I would unconditionally call `this.configureObservable`.

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,210 @@
+    autoInit: true,
+    get storage() {
+      return jsonFileSyncStorageAdapter;
+    },
```
We shouldn't have a default storage adapter here as it doesn't actually track the correct type of `T`

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,210 @@
-    if (this.options.addKeyPrefix) {
-      this.key = StorageHelper.keyPrefix + key;
+export class StorageHelper<T> {
+  static defaultOptions: StorageHelperOptions<any> = {
```
Default options should be untyped (see bellow for reason).

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,210 @@
-  }
+  };
+
+  private data = observable.box<T>();
```
Given that we have default options, we should type them as fully supplied. Furthermore, there shouldn't be a need for an outside inspector (except in tests where we can cast to any anyway) to view the options after construction.
```suggestion

  private options: Required<StorageHelperOptions<T>>;
  private data = observable.box<T>();
```

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,210 @@
-    useSession: false,
-  };
+export interface StorageConfiguration<T> {
+  storage?: StorageAdapter<T>;
```
Given that the default storage adaptor in this PR doesn't cover all `T` this field should be made required. (I don't think that is a breaking change since we don't export `StorageHelper`).

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,210 @@
 
-  protected get storage() {
-    if (this.options.useSession) return window.sessionStorage;
+  constructor(readonly key: string, readonly defaultValue?: T, readonly options: StorageHelperOptions<T> = {}) {
```
Given that options now has a required field it should be a required argument.

```suggestion
  constructor(readonly key: string, options: StorageHelperOptions<T>, readonly defaultValue?: T) {
```

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,210 @@
+  }
+};
+
+// TODO: remove after merge https://github.com/lensapp/lens/pull/2269
```
I assume this is the PR this is supposed to relate to?

```suggestion
// TODO: remove after merge https://github.com/lensapp/lens/pull/2280
```

ixrock replied:
Doesn't really matter since the line cleaned up in follow-up PR (#2280).

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,210 @@
+  initialized: false,
+
+  get filePath() {
+    const clusterId = location.hostname.split(".").slice(-2, -1)[0];
```
This should be gotten from `getHostedClusterId()`

ixrock replied:
Yes, but no.. This is `renderer/utils/*` which should be with minimal dependencies, especially from `main` module.

Nokel81 replied:
I guess I didn't realize that `getHostedClusterId()` was in `main`. That seems like a strange place to have it.

ixrock replied:
Technically it's in `src/common/cluster-store.ts` but I consider it more as `main` stuff which should be avoided in `renderer`. After moving `getHostedClusterId` to appropriate place I'll update it here.

Nokel81 replied:
How is it `main` then? It is used by `getHostedCluster()` all the time.

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,210 @@
+
+    return this.data.get(key);
+  },
+  setItem(key: string, value: any) {
```
We shouldn't have this sort of coercion in a storage abstraction.
```suggestion
  setItem(key: string, value: object) {
```

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,210 @@
-  restoreDefaultValue() {
-    return this.set(this.defaultValue);
+  merge(value: Partial<T> | ((draft: Draft<T>) => Partial<T> | void)) {
+    const updater = isFunction(value) ? value : (state: Draft<T>) => {
```
Is there a reason why you don't use `typeof`? We aren't targeting Safari 9 so I don't see the benefit of calling lodash's function here

ixrock replied:
Because I use already `lodash` in that module, otherwise I prefer `typeof`

ixrock committed:
Refactoring usages of renderer/utils/createStorage (#2280)

* - refactoring: usages of `renderer/utils/createStorage`
- refactoring: Sidebar, SidebarItem and MainLayout
- ux-fix: expanding sidebar by clicking to the arrow-icon (for consistency with compact mode where no sub-menus)

Signed-off-by: Roman <ixrock@gmail.com>

* refactoring, adding some unit tests

Signed-off-by: Roman <ixrock@gmail.com>

* mark StorageHelper.options as private

Signed-off-by: Roman <ixrock@gmail.com>

Nokel81 commented on src/renderer/components/+namespaces/namespace.store.ts:
```diff
@@ -5,14 +5,14 @@ import { Namespace, namespacesApi } from "../../api/endpoints/namespaces.api";
   multiValues: true,
   get defaultValue() {
-    return storage.get() ?? []; // initial namespaces coming from URL or local-storage (default)
+    return selectedNamespaces.get() ?? []; // initial namespaces coming from URL or local-storage (default)
```
Why not pass `[]` in as the default value?

Nokel81 commented on src/renderer/components/dock/dock-tab.store.ts:
```diff
@@ -1,25 +1,30 @@
 interface Options<T = any> {
-  storageName?: string; // name to sync data with localStorage
-  storageSerializer?: (data: T) => Partial<T>; // allow to customize data before saving to localStorage
+  storageName?: string; // persistent key
```
The key isn't persistent but the data behind the key

```suggestion
  storageName?: string; // persistence key
```

Nokel81 commented on src/renderer/components/dock/dock-tab.store.ts:
```diff
@@ -1,25 +1,30 @@
-  storageName?: string; // name to sync data with localStorage
-  storageSerializer?: (data: T) => Partial<T>; // allow to customize data before saving to localStorage
+  storageName?: string; // persistent key
+  storageSerializer?: (data: T) => Partial<T>; // allow to customize data before saving
```
We aren't doing serialization here.

```suggestion
  storageFinalizer?: (data: T) => Partial<T>; // allow to customize data before saving
```

ixrock replied:
Out of PR's scope

Nokel81 replied:
I acknowledge that switching to `Set` and `Map` is out of scope, but why is this out of scope?

ixrock replied:
To not postpone **4.2.0**, all these suggestions are non-critical.

Nokel81 replied:
Why is allowing incorrect names through in a PR postponing the milestone?

ixrock replied:
I have nothing to say anymore. Too many _why_..
Only bugs/critical issues I'm up to discuss and change for this PR.

Nokel81 replied:
That is not how PRs should work. This current name is highly misleading and we shouldn't not be restrict what we can talk about because a release is coming up. If something isn't good enough then we will just bump it to the next release.

jim-docker replied:
If the name is more appropriate it's an easy change, right? This function is not exported to extensions?

ixrock replied:
This functionality moved out from class initialization `options` to `protected` class method (overwritten in `edit-resource.store.ts`)

Nokel81 commented on src/renderer/components/dock/dock-tab.store.ts:
```diff
@@ -1,25 +1,30 @@
 @autobind()
 export class DockTabStore<T = any> {
-  protected data = observable.map<TabId, T>([]);
+  private storage?: StorageHelper<Record<TabId, T>>;
```
Do we even need to have this as a field? It seems like it gets created and then can be handed off to the reaction.

ixrock replied:
There is a missing `this.storage?.clear()` in `clear()`, will be added.

Nokel81 replied:
okay sounds good.

Nokel81 commented on src/renderer/components/dock/dock-tab.store.ts:
```diff
@@ -34,14 +39,19 @@ export class DockTabStore<T = any> {
   }
 
-  protected serializeData() {
+  protected serializeData(): Record<TabId, T> {
```
There is no grantee given the implementation that the values will be complete. We shouldn't mislead future maintainers.

```suggestion
  protected finalizeData(): Record<TabId, Partial<T>> {
```

Nokel81 commented on src/renderer/components/dock/dock.store.ts:
```diff
@@ -21,64 +21,87 @@ export interface IDockTab {
-    { id: "terminal", kind: TabKind.TERMINAL, title: "Terminal" },
-  ];
+export class DockStore implements DockStorageState {
+  readonly minHeight = 100;
```
```suggestion
  static private readonly MinHeight = 100;
```

ixrock committed:
integration tests: refactoring & fixes

Signed-off-by: Roman <ixrock@gmail.com>

Nokel81 commented on src/renderer/components/dock/dock.store.ts:
```diff
@@ -21,64 +21,87 @@ export interface IDockTab {
-  public readonly minHeight = 100;
+  private storage = createStorage<DockStorageState>("dock", {
+    height: 300,
+    tabs: [
+      { id: "terminal", kind: TabKind.TERMINAL, title: "Terminal" },
+    ],
```
Why not have tabs be a `Map<string, DockTab>`? That way we don't need to do all those iterations.

ixrock replied:
It sticks to more plain json-format (this is out of scope of PR) / could be improved later (_dock.store.ts_)

Nokel81 replied:
okay sounds good

Nokel81 commented on src/renderer/components/layout/sidebar-storage.ts:
```diff
@@ -0,0 +1,15 @@
+export interface SidebarStorageState {
+  width: number;
+  compact: boolean;
+  expanded: {
+    [itemId: string]: boolean;
+  }
```
Better describes what we want

```suggestion
  expanded: Set<string>,
```

ixrock replied:
Let's not postpone milestone with such micro optimisations? (could be improved later on demand / follow-up PRs)

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,211 @@
-  return new StorageHelper(key, defaultValue, options);
+export type StorageModel = Record<string, any>; // any json-compatible model
+
+export interface StorageAdapter<T = StorageModel> {
```
This doesn't really do what I think you mean. Plus I don't think it should matter to the consumers how to data is serialized.
```suggestion
export interface StorageAdapter<T> {
```

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,211 @@
+ * Keep intended window.localStorage state in external JSON-file.
+ * Reason: app creates random ports between restarts and as a result storage not persistent.
+ */
+export const jsonFileStorageAdapter: StorageAdapter = {
```
I think that this should be a function that takes validators for the deserialization step (from a defence in depth pov).

Furthermore `any` isn't necessarily JSON serializable so I think a better default would be to use a different format. If the data was sent to `main` and then serialized there we could use `v8.serialize` and get most of this for "free". The other option would be to have our own serialization function that intercepts some types and encodes them in a strict sense.

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,211 @@
+
+    if (data) {
+      this.data.replace(data);
+    }
```
```suggestion
    try {
      const data = await readJson(this.filePath);

      this.data.replace(data);
    } catch () {}
```

ixrock committed:
added missing storage clean-up

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
reverting integration tests refactoring / fixing tests to current branch

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
clean-up: namespace-store, updating comments

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix tests harder

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
reverted: expanding sub-menus by clicking whole sidebar-item area instead of arrow-icon only

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fix integration tests, clean up

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
fixing tests harder, part 2

Signed-off-by: Roman <ixrock@gmail.com>

jim-docker commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,211 @@
+  data: observable.map<string, any>([], { deep: false }),
+  initialized: false,
+
+  get filePath() {
```
Not sure where it would be done but these files (for a cluster) should be deleted when the cluster is removed from Lens. Especially since the same cluster gets a new id every time it is added to Lens.

ixrock replied:
Yes, those are should be deleted, will take a look.

ixrock replied:
Fixed

jim-docker commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,211 @@
+    if (data) {
+      this.data.replace(data);
+    }
+    this.bindAutoSave();
```
this should get disposed when a cluster is removed?

ixrock replied:
Not sure, since everything wiping out from "cluster-scope" including `iframe` (where this code actually running).

jim-docker reviewed (commented):
Looks good but I'd like to test a bit more tomorrow. I noticed my selected namespaces got reset to `[]` when connecting to a flaky kind cluster. Not sure where the flakiness is coming from (seen it on master too, with minikube as well)

ixrock committed:
refactoring, fixes, removing storage json file on cluster removal

Signed-off-by: Roman <ixrock@gmail.com>

jim-docker reviewed (dismissed):
Nice refactor!

ixrock committed:
minor refactoring, added more unit-tests

Signed-off-by: Roman <ixrock@gmail.com>

Nokel81 commented on src/renderer/components/dock/dock-tab.store.ts:
```diff
@@ -1,25 +1,30 @@
 
   constructor(protected options: Options<T> = {}) {
-    const { storageName } = options;
+    this.init();
```
Please catch any possible error here.

ixrock replied:
Refactored a bit, PTAL if still needs extra `try-catch`

Nokel81 commented on src/renderer/components/dock/dock-tab.store.ts:
```diff
@@ -34,14 +39,19 @@ export class DockTabStore<T = any> {
-      if (storageSerializer) return [tabId, storageSerializer(tabData)];
+    if (storageSerializer) {
+      return Object.entries(data).reduce((data, [tabId, tabData]) => {
+        data[tabId] = storageSerializer(tabData) as T;
```
this cast is no longer necessary

ixrock replied:
Refactored, PTAL

Nokel81 commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,70 @@
+    reaction(() => storage.toJSON(), saveFile, { delay: 250 });
+
+    // remove json-file when cluster deleted
+    when(() => clusterStore.removedClusters.has(clusterId)).then(removeFile);
```
We shouldn't have this call if `clusterId` is undefined

ixrock replied:
Fixed

ixrock committed:
fix: show and allow to expand crd sub-menus / sidebar refactoring

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
responding to comments, dock-tab.store.ts refactoring

Signed-off-by: Roman <ixrock@gmail.com>

Nokel81 commented on src/renderer/components/dock/dock-tab.store.ts:
```diff
@@ -34,14 +49,18 @@ export class DockTabStore<T = any> {
+      data[tabId] = this.serializeBeforeSave(tabData);
+
+      return data;
+    }, allTabsData);
```
I would recommend recreating it.
```suggestion
    return Object.fromEntries(
      Object.entries(allTabsData)
        .map(([tabId, tabData]) => ([
          tabId,
          this.finalizeDataForSave(tabData),
        ])
    );
```

Nokel81 commented on src/renderer/components/dock/dock-tab.store.ts:
```diff
@@ -34,14 +49,18 @@ export class DockTabStore<T = any> {
 
-  protected serializeData() {
-    const { storageSerializer } = this.options;
+  protected serializeBeforeSave(data: T): T {
```
Why do you insist that this is called "serialize"? This function isn't doing the following (from wikipedia)

>  translating a data structure or object state into a format that can be stored

That would be a call to `JSON.serialize` or similar.

ixrock replied:
Ok, I'll update the method name

Nokel81 reviewed (commented):
one last thing

ixrock committed:
responding to comments, dock-tab.store more tweaks

Signed-off-by: Roman <ixrock@gmail.com>

ixrock committed:
crd listing fixes

Signed-off-by: Roman <ixrock@gmail.com>

ixrock commented on src/renderer/components/+custom-resources/crd-list.tsx:
```diff
@@ -60,13 +68,9 @@ export class CrdList extends React.Component {
+        items={items}
         sortingCallbacks={sortingCallbacks}
         searchFilters={Object.values(sortingCallbacks)}
-        filterItems={[
```
Btw, this filtering seems to be broken in `4.1.4`

Nokel81 replied:
Broken how, like it isn't being called at all? That is not good since we have to keep that prop working as it is exported.

aleksfront commented on src/renderer/utils/createStorage.ts:
```diff
@@ -1,73 +1,72 @@
-      } catch (e) {
-        console.error(`Parsing json failed for pair: ${this.key}=${strValue}`);
-      }
+  async function saveFile(json = {}) {
```
Maybe `saveToAppFiles()` could be used instead? `src/common/utils/saveToAppFiles.ts`

ixrock replied:
Maybe, but it uses `fs-extra` sync file operations, e.g. `fse.writeFileSync`

aleksfront reviewed (approved):
Work as expected.

Nokel81 reviewed (approved):
Please leave merging to me, I need to coordinate it with the rc.1 PR
